libknet/handle.c

```diff
@@ -1482,42 +1489,111 @@ int knet_handle_crypto(knet_handle_t knet_h, struct knet_handle_crypto_cfg *knet
 		return -1;
 	}
 
-	if ((!strncmp("none", knet_handle_crypto_cfg->crypto_model, 4)) || 
+	if (knet_h->crypto_in_use_config == config_num) {
+		savederrno = EBUSY;
+		err = -1;
+		goto exit_unlock;
+	}
+
+	if ((!strncmp("none", knet_handle_crypto_cfg->crypto_model, 4)) ||
 	    ((!strncmp("none", knet_handle_crypto_cfg->crypto_cipher_type, 4)) &&
 	     (!strncmp("none", knet_handle_crypto_cfg->crypto_hash_type, 4)))) {
-		crypto_fini(knet_h);
-		log_debug(knet_h, KNET_SUB_CRYPTO, "crypto is not enabled");
+		crypto_fini(knet_h, config_num);
+		log_debug(knet_h, KNET_SUB_CRYPTO, "crypto config %u is not enabled", config_num);
 		err = 0;
 		goto exit_unlock;
 	}
 
 	if (knet_handle_crypto_cfg->private_key_len < KNET_MIN_KEY_LEN) {
-		log_debug(knet_h, KNET_SUB_CRYPTO, "private key len too short (min %d): %u",
-			  KNET_MIN_KEY_LEN, knet_handle_crypto_cfg->private_key_len);
+		log_debug(knet_h, KNET_SUB_CRYPTO, "private key len too short for config %u (min %d): %u",
+			  config_num, KNET_MIN_KEY_LEN, knet_handle_crypto_cfg->private_key_len);
 		savederrno = EINVAL;
 		err = -1;
 		goto exit_unlock;
 	}
 
 	if (knet_handle_crypto_cfg->private_key_len > KNET_MAX_KEY_LEN) {
-		log_debug(knet_h, KNET_SUB_CRYPTO, "private key len too long (max %d): %u",
-			  KNET_MAX_KEY_LEN, knet_handle_crypto_cfg->private_key_len);
+		log_debug(knet_h, KNET_SUB_CRYPTO, "private key len too long for config %u (max %d): %u",
+			  config_num, KNET_MAX_KEY_LEN, knet_handle_crypto_cfg->private_key_len);
 		savederrno = EINVAL;
 		err = -1;
 		goto exit_unlock;
 	}
 
-	err = crypto_init(knet_h, knet_handle_crypto_cfg);
+	err = crypto_init(knet_h, knet_handle_crypto_cfg, config_num);
 
 	if (err) {
 		err = -2;
 		savederrno = errno;
 	}
 
 exit_unlock:
-	if (!err) {
-		force_pmtud_run(knet_h, KNET_SUB_CRYPTO, 1);
+	pthread_rwlock_unlock(&knet_h->global_rwlock);
+	errno = err ? savederrno : 0;
+	return err;
+}
+
+int knet_handle_crypto_rx_clear_traffic(knet_handle_t knet_h,
+					uint8_t value)
+{
+	int savederrno = 0;
+
+	if (!knet_h) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (value > 1) {
```

Comment:
I think it would be cleaner to use KNET_CRYPTO_RX_DISALLOW_CLEAR_TRAFFIC here rather than '1' or maybe even have a _MAX #define for it.